### DISCUSSION TITLE
Trying to understand where `islandsByPath` are filled

### DISCUSSION THREAD

**Discussion by codepunkt** [ElMassimo/iles#171] [Q&A] [answered]
I'm trying to understand where `islandsByPath` gets filled. When debugging, it seems like the [call to `renderToString`](https://github.com/ElMassimo/iles/blob/main/packages/iles/src/node/build/render.ts#L48) from `@vue/server-renderer` adds islands for the rendered path to the object by invoking `renderToString` with `{ islandsByPath, renderers }` as second `SSRContext` parameter, but I'm completely dumbfounded on how this is done and why this works.

**Comment by ElMassimo** [comment, marked as answer]
Interesting question 😃 

The answer is, through [Vue's SSR Context](https://vuejs.org/api/ssr.html#rendertostring).

You can see it being used in [`useIslandsForPath`](https://github.com/ElMassimo/iles/blob/improve/vue-hydration/packages/iles/src/client/app/composables/islandDefinitions.ts#L6-L14).

This composable is [used in the `Island` component](https://github.com/ElMassimo/iles/blob/improve/vue-hydration/packages/iles/src/client/app/components/Island.vue#L108) to gather information about rendered islands, in order to allow Rollup to build an optimized bundle for any JavaScript that should be shipped in the site.

**Reply by codepunkt** [reply]
That's a great answer, thanks! Where is the `Island` component being used? I couldn't find any reference to it.

**Reply by ElMassimo** [reply]
It's automatically [injected](https://github.com/ElMassimo/iles/blob/improve/vue-hydration/packages/iles/src/node/plugin/remarkWrapIslands.ts#L62) when a [hydration directive](https://iles-docs.netlify.app/guide/hydration#hydration-directives) is [detected](https://github.com/ElMassimo/iles/blob/improve/vue-hydration/packages/iles/src/node/plugin/wrap.ts#L84-L87), both in MDX and Vue templates.

Code is transformed using [Vite plugins](https://github.com/ElMassimo/iles/blob/improve/vue-hydration/packages/iles/src/node/plugin/plugin.ts#L122-L134).

Add [`vite-plugin-inspect`](https://github.com/antfu/vite-plugin-inspect) to a site, and visit the inspector for a Vue file with a hydration directive, and you will be able to see the transformations step-by-step.

**Reply by codepunkt** [reply]
That's genius! I'm trying to build something similar, but for `preact` only. Seems a lot harder, as it doesn't have the parser, amongst other things.

**Reply by ElMassimo** [reply]
You might be interested on [this framework](https://github.com/bensmithett/tropical) which uses React.

The reason I picked Vue is that plain HTML is valid as a Vue template, and Vue is great at encapsulation while [being convenient](https://vuejs.org/guide/components/attrs.html#fallthrough-attributes), making it ideal for both dynamic and static sites.

Just in case, have in mind that you __can__ use [Preact islands](https://iles.pages.dev/guide/frameworks) in îles, mentioning it because some people assume it's only for Vue.
